### ISSUE TITLE
Training BERT dinamically with features and on non redundant data 

### ISSUE THREAD

**Issue by Bedinka** [Bedinka/PPI-prediction-with-Deep-Learning-Language-Model-#11]
# PPI Analysis with Deep Learning Language Model Project

## Description

Training BERT on cathegorized and numerical datat beside the sequencial input. 
Based on the work of: https://mccormickml.com/2021/06/29/combining-categorical-numerical-features-with-bert/
Augmented BERt is used for binary classification prediction of the _label : Interact_ in the dataset. 

### The changes to the file can be found [combine_input_bert.py](https://github.com/lionking0000/YangLabIntern/blob/master/PPI_W_DLLM/combine_input_bert.py).

---

### Problem Statement

BERT is a Language model and accepts strings as an input. So it was needed to augment it so it can recove numerical and cathegorical data at the same time for the training. 

---

### Proposed Solution

_Augmented BERT solution:_

- Adding each information from teh tsv column into a sequence of string with {:}[SEP]{:}[SEP]{:}".format . 
- Adding each attribute dinamically. No need to adjust int the BERT code just the call of the BERT file. 
- Saving each BETT model 
- Creating a summary tsv from the run to determine the BEST perfoming BERT.

_Testing attributes:_
It was neede to evaluate how each attribute added for the training data contributes to the prediction. I counducted the F1 score to validate it. Training different BERT models bsed on haw many attributes are found in the training tsv. The attrnum in the Bert models name coppespond to how many were used for the training. 
Used attributes : 
- Sequence 
- DSSP index 
- DSSP struct 
- Vector 

---

### Additional Context
[Add any additional context, screenshots, or information that may be relevant to the issue.]

---

### Related Code

_Dinamical input loading :_ 

    print('Combining features into strings...')

    # For each of the samples...
    for index, row in data_df.iterrows():
        fields = [str(row[field]) for field in combined_fields]
        combined = '[SEP]'.join(fields)
        #print(combined)

        sen_w_feats.append(combined)
        labels.append(row["Interact"])

 OLD VERSION: not dinamic 
        """# Piece it together...    
        combined = ""
        #combined += "The ID of this item is {:}, ".format(row["Clothing ID"])
        combined += "{:}[SEP]{:}[SEP]{:}".format(row["Protein ID"],
                                                                row["Residues"],
                                                                row["DSSP Structure"])"""

---

### Results

**_Testing BERT results:_** 

atr_1: 
![bert_attrnum_1_no_500_s_wDataloader](https://github.com/lionking0000/YangLabIntern/assets/105316730/e043a46a-7e74-4be1-9a16-dcb4276c3d18)
atr_2:
![bert_attrnum_2_no_500_s_wDataloader](https://github.com/lionking0000/YangLabIntern/assets/105316730/2a36aa01-41c7-42f2-b089-eb65a8e4c44b)
atr_3:
![bert_attrnum_3_no_500_s_wDataloader](https://github.com/lionking0000/YangLabIntern/assets/105316730/0e480eb0-cb1f-456d-aed6-f0827e2102e0)
atr_4:
![bert_attrnum_4_no_500_s_wDataloader](https://github.com/lionking0000/YangLabIntern/assets/105316730/c4299813-075f-4740-a52e-e83be4b222ba)

**_Training intermadiate results:_**

<!DOCTYPE html>

<html>
<head>
	
	
	<title></title>
	<meta name="generator" content="LibreOffice 7.3.7.2 (Linux)"/>
	<style type="text/css">
		body,div,table,thead,tbody,tfoot,tr,th,td,p { font-family:"Liberation Sans"; font-size:x-small }
		a.comment-indicator:hover + comment { background:#ffd; position:absolute; display:block; border:1px solid black; padding:0.5em;  } 
		a.comment-indicator { background:red; display:inline-block; border:1px solid black; width:0.5em; height:0.5em;  } 
		comment { display:none;  } 
	</style>
	
</head>

<body>

Training Loss | Valid. Loss | Valid. Accur. | Training Time | Validation Time | F1 Score | model_name
-- | -- | -- | -- | -- | -- | --
0.685459343643932 | 0.720131856203079 | 0.5225 | 0:00:06 | 0:00:00 | 0.675675675675676 |  
0.670817111219679 | 0.737256193161011 | 0.5225 | 0:00:06 | 0:00:00 | 0.675675675675676 |  
0.665647203659082 | 0.7547572016716 | 0.4725 | 0:00:06 | 0:00:00 | 0.675675675675676 |  
0.646912665723206 | 0.742187750339508 | 0.585 | 0:00:06 | 0:00:00 | 0.675675675675676 |  
0.616083348726297 | 0.742826592922211 | 0.5725 | 0:00:06 | 0:00:00 | 0.675675675675676 |  
0.619144921178942 | 0.754417502880096 | 0.5525 | 0:00:06 | 0:00:00 | 0.675675675675676 |  
0.562894591650405 | 0.776118993759155 | 0.5475 | 0:00:06 | 0:00:00 | 0.675675675675676 |  
0.545238047451168 | 0.802140247821808 | 0.5775 | 0:00:06 | 0:00:00 | 0.675675675675676 |  
0.5263597144709 | 0.76742345392704 | 0.59 | 0:00:06 | 0:00:00 | 0.675675675675676 |  
0.496193420190316 | 0.763749504089356 | 0.59 | 0:00:06 | 0:00:00 | 0.675675675675676 |  
  |   |   |   |   |   | bert_attrnum_1_no_500_s_wDataloader
0.719376978162047 | 0.665669798851013 | 0.625 | 0:00:27 | 0:00:01 | 0.804878048780488 |  
0.685869817222868 | 0.602256560325623 | 0.7 | 0:00:27 | 0:00:01 | 0.804878048780488 |  
0.637663505294106 | 0.571521186828613 | 0.685 | 0:00:27 | 0:00:01 | 0.804878048780488 |  
0.621652907365328 | 0.544720080494881 | 0.7875 | 0:00:27 | 0:00:01 | 0.804878048780488 |  
0.610269729967241 | 0.494001230597496 | 0.7875 | 0:00:27 | 0:00:01 | 0.804878048780488 |  
0.579108768469327 | 0.493873247504234 | 0.7875 | 0:00:27 | 0:00:01 | 0.804878048780488 |  
0.555797838351943 | 0.460552714765072 | 0.7675 | 0:00:27 | 0:00:01 | 0.804878048780488 |  
0.500670782931439 | 0.467086625099182 | 0.7675 | 0:00:27 | 0:00:01 | 0.804878048780488 |  
0.491499402499818 | 0.434261159598827 | 0.85 | 0:00:27 | 0:00:01 | 0.804878048780488 |  
0.465762782406497 | 0.437377798557282 | 0.8375 | 0:00:27 | 0:00:01 | 0.804878048780488 |  
  |   |   |   |   |   | bert_attrnum_2_no_500_s_wDataloader
0.713637013714035 | 0.671404606103897 | 0.615 | 0:00:27 | 0:00:01 | 0.784810126582279 |  
0.69196216162149 | 0.638651758432388 | 0.6125 | 0:00:27 | 0:00:01 | 0.784810126582279 |  
0.669944715577287 | 0.552972173690796 | 0.705 | 0:00:27 | 0:00:01 | 0.784810126582279 |  
0.634644075260534 | 0.533533982932568 | 0.7375 | 0:00:27 | 0:00:01 | 0.784810126582279 |  
0.602064236417993 | 0.608842620253563 | 0.6625 | 0:00:27 | 0:00:01 | 0.784810126582279 |  
0.557778713571561 | 0.576527464389801 | 0.7175 | 0:00:27 | 0:00:01 | 0.784810126582279 |  
0.533049204132774 | 0.538173133134842 | 0.755 | 0:00:27 | 0:00:01 | 0.784810126582279 |  
0.481241980156341 | 0.522353568673134 | 0.7425 | 0:00:27 | 0:00:01 | 0.784810126582279 |  
0.454782143041685 | 0.591610550880432 | 0.7 | 0:00:27 | 0:00:01 | 0.784810126582279 |  
0.442037726958077 | 0.559501096606255 | 0.725 | 0:00:27 | 0:00:01 | 0.784810126582279 |  
  |   |   |   |   |   | bert_attrnum_3_no_500_s_wDataloader
0.708531893693007 | 0.623913750052452 | 0.7 | 0:00:27 | 0:00:01 | 0.763157894736842 |  
0.681494162841277 | 0.595031362771988 | 0.7125 | 0:00:27 | 0:00:01 | 0.763157894736842 |  
0.648826416823771 | 0.597771495580673 | 0.6625 | 0:00:27 | 0:00:01 | 0.763157894736842 |  
0.628761639068653 | 0.582911437749863 | 0.7125 | 0:00:27 | 0:00:01 | 0.763157894736842 |  
0.618247043002735 | 0.6186817497015 | 0.6925 | 0:00:27 | 0:00:01 | 0.763157894736842 |  
0.571508251227342 | 0.600395891070366 | 0.7 | 0:00:27 | 0:00:01 | 0.763157894736842 |  
0.539209706055654 | 0.588294717669487 | 0.73 | 0:00:27 | 0:00:01 | 0.763157894736842 |  
0.528535446950367 | 0.599666818976402 | 0.73 | 0:00:27 | 0:00:01 | 0.763157894736842 |  
0.481118611701123 | 0.629218232631683 | 0.6675 | 0:00:27 | 0:00:01 | 0.763157894736842 |  
0.44994115732707 | 0.612168362736702 | 0.6675 | 0:00:27 | 0:00:01 | 0.763157894736842 |  
  |   |   |   |   |   | bert_attrnum_4_no_500_s_wDataloader

</body>

</html>

**_Log of the Training :_**

**2024-06-05 19:03:41,961 - INFO - Running BERT model bert_attrnum_1_no_500_s_wDataloader with combined fields: ['Residues']**
2024-06-05 19:03:41,972 - INFO -   Training size: 616
2024-06-05 19:03:41,972 - INFO - Validation size: 77
2024-06-05 19:03:41,972 - INFO -       Test size: 77
2024-06-05 19:03:41,972 - INFO - Combining features into strings...
2024-06-05 19:03:41,994 - INFO -   DONE.
2024-06-05 19:03:41,994 - INFO - Dataset contains 770 samples.
2024-06-05 19:03:42,045 - INFO - Loading BERT tokenizer...
2024-06-05 19:04:48,340 - INFO - Model saved as bert_attrnum_1_no_500_s_wDataloader in directory ./BERT/bert_attrnum_1_no_500_s_wDataloader
2024-06-05 19:04:48,466 - INFO - Predicting labels for 77 test sequences...
2024-06-05 19:04:48,640 - INFO -     DONE.
**2024-06-05 19:04:48,642 - INFO - F1 Score: 0.676**
2024-06-05 19:04:48,646 - INFO - ----------------------------------------------------------------------------------------------------
**2024-06-05 19:04:48,646 - INFO - Running BERT model bert_attrnum_2_no_500_s_wDataloader with combined fields: ['Residues', 'DSSP Structure']**
2024-06-05 19:04:48,655 - INFO -   Training size: 616
2024-06-05 19:04:48,656 - INFO - Validation size: 77
2024-06-05 19:04:48,656 - INFO -       Test size: 77
2024-06-05 19:04:48,656 - INFO - Combining features into strings...
2024-06-05 19:04:48,677 - INFO -   DONE.
2024-06-05 19:04:48,677 - INFO - Dataset contains 770 samples.
2024-06-05 19:04:48,677 - INFO - Loading BERT tokenizer...
2024-06-05 19:09:36,216 - INFO - Model saved as bert_attrnum_2_no_500_s_wDataloader in directory ./BERT/bert_attrnum_2_no_500_s_wDataloader
2024-06-05 19:09:36,282 - INFO - Predicting labels for 77 test sequences...
2024-06-05 19:09:37,347 - INFO -     DONE.
**2024-06-05 19:09:37,348 - INFO - F1 Score: 0.805**
2024-06-05 19:09:37,352 - INFO - ----------------------------------------------------------------------------------------------------
**2024-06-05 19:09:37,352 - INFO - Running BERT model bert_attrnum_3_no_500_s_wDataloader with combined fields: ['Residues', 'DSSP Structure', 'DSSP Index']**
2024-06-05 19:09:37,362 - INFO -   Training size: 616
2024-06-05 19:09:37,362 - INFO - Validation size: 77
2024-06-05 19:09:37,362 - INFO -       Test size: 77
2024-06-05 19:09:37,362 - INFO - Combining features into strings...
2024-06-05 19:09:37,385 - INFO -   DONE.
2024-06-05 19:09:37,385 - INFO - Dataset contains 770 samples.
2024-06-05 19:09:37,385 - INFO - Loading BERT tokenizer...
2024-06-05 19:14:27,547 - INFO - Model saved as bert_attrnum_3_no_500_s_wDataloader in directory ./BERT/bert_attrnum_3_no_500_s_wDataloader
2024-06-05 19:14:27,617 - INFO - Predicting labels for 77 test sequences...
2024-06-05 19:14:28,684 - INFO -     DONE.
**2024-06-05 19:14:28,685 - INFO - F1 Score: 0.785**
2024-06-05 19:14:28,689 - INFO - ----------------------------------------------------------------------------------------------------
**2024-06-05 19:14:28,689 - INFO - Running BERT model bert_attrnum_4_no_500_s_wDataloader with combined fields: ['Residues', 'DSSP Structure', 'DSSP Index', 'Vector']**
2024-06-05 19:14:28,698 - INFO -   Training size: 616
2024-06-05 19:14:28,699 - INFO - Validation size: 77
2024-06-05 19:14:28,699 - INFO -       Test size: 77
2024-06-05 19:14:28,699 - INFO - Combining features into strings...
2024-06-05 19:14:28,722 - INFO -   DONE.
2024-06-05 19:14:28,722 - INFO - Dataset contains 770 samples.
2024-06-05 19:14:28,722 - INFO - Loading BERT tokenizer...
2024-06-05 19:14:33,580 - INFO - Max sentence length: 
2024-06-05 19:14:33,580 - INFO - 512
2024-06-05 19:19:21,224 - INFO - Model saved as bert_attrnum_4_no_500_s_wDataloader in directory ./BERT/bert_attrnum_4_no_500_s_wDataloader
2024-06-05 19:19:21,306 - INFO - Predicting labels for 77 test sequences...
2024-06-05 19:19:22,397 - INFO -     DONE.
**2024-06-05 19:19:22,398 - INFO - F1 Score: 0.763**
2024-06-05 19:19:22,403 - INFO - BERT training stats saved to BERT_training_stats_withvectors_2024-06-05_17-17-39_500_s_wDataloader.tsv

---
### Checklist for coding 
- [x] Modul configuration.
- [x] Augmentation non dinemically. 
- [x] Augmentation dinamically .
- [x] Testing each attribute.
- [x] Defining the best attributes for training

### Checklist for tasks 
- [x] Reviewed the problem statement and proposed solution.
- [x] Environmental dependecy solutions.
- [x] Provided expected and actual behavior matches.
- [x] Included all relevant context and code snippets.
- [x] Commited.
